metadata-ingestion/src/datahub/utilities/sqlalchemy_query_combiner.py

```diff
@@ -0,0 +1,345 @@
+import collections
+import contextlib
+import dataclasses
+import itertools
+import logging
+import random
+import string
+import threading
+import unittest.mock
+from typing import Any, Callable, Dict, Iterator, List, Optional, Set, Tuple
+
+import greenlet
+import sqlalchemy
+import sqlalchemy.engine
+import sqlalchemy.sql
+from sqlalchemy.engine import Connection
+from sqlalchemy.orm.exc import MultipleResultsFound, NoResultFound
+from typing_extensions import ParamSpec
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+P = ParamSpec("P")
+
+
+# We need to make sure that only one query combiner attempts to patch
+# the SQLAlchemy execute method at a time so that they don't interfere.
+# Generally speaking, there will only be one query combiner in existence
+# at a time anyways, so this lock shouldn't really be doing much.
+_sa_execute_method_patching_lock = threading.Lock()
+_sa_execute_underlying_method = sqlalchemy.engine.Connection.execute
+
+
+def _generate_sql_safe_identifier() -> str:
```

Comment:
should this be local to _handle_execute or a static method of `SQLAlchemyQueryCombiner`

Reply:
It can be a static method - will change!